// vscode-tmgrammar-test/dist/src/snapshot/parsing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function parseSnap(s) {
    var result = [];
    var ls = s.split(/\r\n|\n/);
    var i = 0;
    while (i < ls.length) {
        var l = ls[i];
        if (l.startsWith('>')) {
            var src = l.substr(1);
            i++;
            var tokens = [];
            while (i < ls.length && ls[i].startsWith('#')) {
                var startIndex = ls[i].indexOf('^');
                var endIndex = ls[i].indexOf(' ', startIndex);
                var scopes = ls[i]
                    .substr(endIndex + 1)
                    .split(/\s+/)
                    .filter(function (x) { return x !== ''; });
                tokens.push({
                    startIndex: startIndex - 1,
                    endIndex: endIndex - 1,
                    scopes: scopes
                });
                i++;
            }
            result.push({
                src: src,
                tokens: tokens
            });
        }
        else {
            i++;
        }
    }
    return result;
}
exports.parseSnap = parseSnap;
function renderSnap(xs) {
    var result = [];
    xs.forEach(function (line) {
        result.push('>' + line.src);
        if (line.src.trim().length > 0) {
            line.tokens.forEach(function (token) {
                result.push('#' +
                    ' '.repeat(token.startIndex) +
                    '^'.repeat(token.endIndex - token.startIndex) +
                    ' ' +
                    token.scopes.join(' '));
            });
        }
    });
    return result.join('\n');
}
exports.renderSnap = renderSnap;
